core/consensus/transport.go

```diff
@@ -0,0 +1,188 @@
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package consensus
+
+import (
+	"context"
+	"sync"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/core"
+	pbv1 "github.com/obolnetwork/charon/core/corepb/v1"
+	"github.com/obolnetwork/charon/core/qbft"
+)
+
+// transport encapsulates receiving and broadcasting for a consensus instance/duty.
+type transport struct {
+	// Immutable state
+	component  *Component
+	recvBuffer chan qbft.Msg[core.Duty, [32]byte] // Instance inner receive buffer.
+
+	// Mutable state
+	valueMu sync.Mutex
+	values  map[[32]byte]*pbv1.UnsignedDataSet // maps proposed values to their hashes
+}
+
+// setValues caches the values and their hashes.
+func (t *transport) setValues(msg msg) {
+	t.valueMu.Lock()
+	defer t.valueMu.Unlock()
+
+	t.values[msg.Value()] = msg.msg.Value
+	t.values[msg.PreparedValue()] = msg.msg.PreparedValue
+}
+
+// getValue returns the value by its hash.
+func (t *transport) getValue(hash [32]byte) (*pbv1.UnsignedDataSet, error) {
+	t.valueMu.Lock()
+	defer t.valueMu.Unlock()
+
+	data, ok := t.values[hash]
+	if !ok {
+		return nil, errors.New("unknown value")
+	}
+
+	return data, nil
+}
+
+// Broadcast creates a msg and sends it to all peers (including self).
+func (t *transport) Broadcast(ctx context.Context, typ qbft.MsgType, duty core.Duty,
+	peerIdx int64, round int64, valueHash [32]byte, pr int64, pvHash [32]byte,
+	justification []qbft.Msg[core.Duty, [32]byte],
+) error {
+	// Get the values by their hashes.
+	value, err := t.getValue(valueHash)
+	if err != nil {
+		return err
+	}
+	pv, err := t.getValue(pvHash)
+	if err != nil {
+		return err
+	}
+
+	// Make the message
+	msg, err := createMsg(typ, duty, peerIdx, round, value, pr, pv, justification)
+	if err != nil {
+		return err
+	}
+
+	// First send to self.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case t.recvBuffer <- msg:
+	}
+
+	for i, p := range t.component.peers {
+		if int64(i) == t.component.peerIdx {
+			// Do not broadcast to self
+			continue
+		}
+
+		err := send(ctx, t.component.tcpNode, p.ID, msg.ToConsensusMsg())
+		if err != nil {
+			log.Warn(ctx, "Failed sending message", err)
+		}
+	}
+
+	return nil
+}
+
+// ProcessReceives processes received messages from the outer buffer until the context is closed.
+func (t *transport) ProcessReceives(ctx context.Context, outerBuffer chan msg) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case msg := <-outerBuffer:
+			if err := validateMsg(msg); err != nil {
+				log.Warn(ctx, "Dropping invalid message", err)
+				continue
+			}
+			t.setValues(msg)
+
+			select {
+			case <-ctx.Done():
```

Comment:
This is in many places so I take it its intentional, but what's with a `select { }` block not counting for an extra indentation, the  `case`s aren't tabbed, convention I guess but seems weird to me.